### ISSUE TITLE
Implement Workspace Switcher

### ISSUE THREAD

**Issue by hayata-suenaga** [Expensify/App#31887]
Display a list of workspaces when the workspace avatar is clicked so that users can "switch" or select workspaces from the list.

<img width="200" alt="Screenshot 2023-11-26 at 9 31 34 AM" src="https://github.com/Expensify/App/assets/98560306/f9f9feef-ccd4-43c5-94ee-91dd5d33c3aa">

The list of workspace should look like this:

<img width="400" alt="Screenshot 2023-11-26 at 9 32 48 AM" src="https://github.com/Expensify/App/assets/98560306/ffdb0d71-e40b-4d52-91ba-677ca7960fb0">

Users use this list of workspaces to switch between workspaces.

More detailed specifications are found in [this section of the design doc](https://docs.google.com/document/d/1VnEf2ThbbiQj0kx9KnAihF0DWf50s2Bn_MztEyEwtjg/edit?pli=1#bookmark=id.sk2sq7wwy4ap).

Note that the Workspace Switcher is different from the list of workspaces that is displayed on [the Settings page of the Expensify/All](https://docs.google.com/document/d/1VnEf2ThbbiQj0kx9KnAihF0DWf50s2Bn_MztEyEwtjg/edit?pli=1#bookmark=id.xk7810ttgzat).

The completion of this issue includes:

- Get all workspaces the user belongs to
- Workspace Search feature
- Highlight the workspace that is currently active with the checkmark
- Display of RBR and GBR depending on if the workspace has any reports with an error or reports that need user attention (ex. money request that hasn't been paid).
- Changing the active workspace when the new workspace is selected from the list.

_The high level implementation is specified in [this part of the design doc](https://docs.google.com/document/d/1VnEf2ThbbiQj0kx9KnAihF0DWf50s2Bn_MztEyEwtjg/edit?pli=1#bookmark=id.lki0mextico0). You don't have to strictly follow the implementation ideas suggested in the design doc or in the above bullet point list_

### Special notes about BRB and GBR requirements
- A utility function is implemented in [this issue](https://github.com/Expensify/App/issues/31878). Assume that this utility function exists.
- Given a workspace, this utility function iterates over all the reports for the given workspace to figure out if it contains any reports with an error or any unread reports.
- The usage of this utility function should be minimized by memoizing the result of the calculation as much as possible.
- Take a special care to figure out RBR and GBR for the Expensify/All workspace (that represents reports across all workspaces). If any reports across workspaces has an error or need attention, RBR or GBR should be displayed accordingly.
- We already know if each workspace contains RBR or GBR. Use that information to figure out BRB or GBR or the Expensify/All workspace. Note that Expensify/All also contains off-workspace reports (reports that don't belong to any workspace). Even if any of the workspaces don't have RBR or GBR, the Expensify/All should have RBR or GBR if any off-workspace reports has an error or unread.

**Comment by trjExpensify** [comment]
Oh cool, but we're held here until the detailed doc is written and through review?

**Comment by hayata-suenaga** [comment]
Yes @trjExpensify I was drafting the issue but didn't know people are automatically assigned

Changed the title to make it clear that no one should start working on this now

**Comment by trjExpensify** [comment]
Doc is in review :)

**Comment by trjExpensify** [comment]
Latest is [here](https://expensify.slack.com/archives/C036QM0SLJK/p1702662895353029). We should have this on `daily` right @hayata-suenaga?

**Comment by roryabraham** [comment]
Not sure if I have all the latest context here, but just FYI I'm making changes across a few PRs to bring an `activePolicyID` field to the account object in Onyx. This is an existing piece of data that's used to determine which policy to attach expenses to when generating expenses via SmartScan.

It will need to be tracked separately from the current active policy in the workspace switcher we're building here. There's some internal slack discussion about this [here](https://expensify.slack.com/archives/C0671JDRKQW/p1702503112841719), but some important notes:

> But it sounds like we’re not just talking about receipt forwarding explicitly in this thread. We’re talking about potentially using the activePolicyID to populate a default workspace in the new room creation flow. So is the working idea this?
> - When in the Expensify “all” view, populate the activePolicyID workspace as the default
> - When “filtered” to a specific workspace view, use the selectedWorkspace as the most contextually relevant workspace to populate when going to create a new room.
> - The “workspace” field in the new room creation flow can be changed in both cases

So basically, `activePolicyID !== selectedPolicy`. The `activePolicyID` is a persistent piece of data that lives in the back-end and is used for SmartScan, among other things. The `selectedPolicy` is something that will be front-end-only and likely introduced in this issue, and it represents the workspace that's currently selected in the workspace switcher.

Hopefully that clarifies things / provides useful context.

**Comment by roryabraham** [comment]
So I think part of the implementation here should be to examine the places where `activePolicyID` is used, and implement the above logic in those places such that when looking to choose a default workspace, we first choose the `selectedPolicy` (i.e: selected in the workspace switcher), and if it's not set (i.e: no workspace selected in the workspace switcher), then fall back on the `activePolicyID`.

**Comment by trjExpensify** [comment]
>such that when looking to choose a default workspace, we first choose the selectedPolicy (i.e: selected in the workspace switcher), and if it's not set (i.e: no workspace selected in the workspace switcher), then fall back on the activePolicyID.

For what feature/purpose exactly though? For example, we don't want to filter the "All" view by the activePolicyID, as that would filter out all chats in the LHN apart from those associated with your "default workspace", which is the opposite of what the workspace switcher feature is designed to do. 

So I think anything that looking to use the activePolicyID (i.e receipt forwarding, workspace "pre-selected" when creating a new room etc) is not in scope of the implementation here for the workspace switcher as part of the ideal nav skeleton project. 

CC: @JmillsExpensify in case I missed some development while out last week.

**Comment by roryabraham** [comment]
Sorry, I'm a bit confused exactly what we're talking about at this point so maybe it's not productive to continue to discuss in the abstract

**Comment by trjExpensify** [comment]
@mateuuszzzzz do you have a WIP PR to link to this issue?

**Comment by hayata-suenaga** [comment]
@mateuuszzzzz are you working on this in the internal branch that Software Mansion created? or do you have a separate PR that you're working on?

**Comment by hayata-suenaga** [comment]
@roryabraham Let me know if we should make the `activeWorkspaceId` the fallback selection when `selectedWorkspaceId` is not available

**Comment by roryabraham** [comment]
@hayata-suenaga I think that makes sense in general but can we ground the conversation with a specific example in-product?

I see what @trjExpensify was saying above – in the context of the workspace switcher I guess that a filter for "all" workspaces would be the default, not the `activeWorkspaceID`, and then when you filter by a given workspace that becomes the `selectedWorkspaceID`.

An example when it would make sense to make the `activeWorkspaceID` the fallback selection when `selectedWorkspaceID` is not available is when pre-populating the workspace in the new room creation flow.

If any usages of `selectedWorkspaceID` for anything other than rendering come up during, let's talk about them here or in slack.

**Comment by trjExpensify** [comment]
Yep, agreed. Good summary. I don't think there's any action here for now! 👍

**Comment by mateuuszzzzz** [comment]
> @mateuuszzzzz are you working on this in the internal branch that Software Mansion created? or do you have a separate PR that you're working on?

Hi, yes I'm working on [internal branch](https://github.com/software-mansion-labs/expensify-app-fork/pull/15). Perhaps you can see the contents of the URL. I mentioned on slack that we need to wait for @adamgrzybowski to integrate this feature with LAST_VISITED_PATH, because it requires changes in current navigation.

**Comment by trjExpensify** [comment]
Work in progress, discussed a good gotcha today [here](https://expensify.slack.com/archives/C036QM0SLJK/p1704975450475039).

**Comment by trjExpensify** [comment]
Finalising that one in thread!

**Comment by hayata-suenaga** [comment]
the latest daily update is [here](https://expensify.slack.com/archives/C036QM0SLJK/p1705428373855649)

**Comment by trjExpensify** [comment]
Adhoc build is looking good, few bugs to resolve are in progress! Latest over [here](https://expensify.slack.com/archives/C036QM0SLJK/p1705684861059829).

**Comment by hayata-suenaga** [comment]
catching up with slack discussions (comment to remove `overdue` for me)

**Comment by hayata-suenaga** [comment]
the feature branch is very close to be merged

**Comment by trjExpensify** [comment]
PR merged today!

**Comment by hayata-suenaga** [comment]
the feature branch was merged. C+ payment is handled in another issue. Closing this issue now
